packages/api/pages/api/auth/[...nextauth].ts

```diff
@@ -26,74 +36,88 @@ const ORIGIN_MATCH = process.env.ORIGIN_MATCH
+  adapter: PrismaAdapter(client as unknown as PrismaClient),
+  providers: [
+    EmailProvider({
+      server: process.env.EMAIL_SERVER,
+      from: process.env.EMAIL_FROM,
+    }),
+  ],
+  cookies: {
+    sessionToken: {
+      name: `${cookiePrefix}next-auth.session-token`,
+      options: secureCookieOptions,
     },
-    cookies: {
-      sessionToken: {
-        name: `${cookiePrefix}next-auth.session-token`,
-        options: secureCookieOptions,
-      },
-      callbackUrl: {
-        name: `${cookiePrefix}next-auth.callback-url`,
-        options: secureCookieOptions,
-      },
-      csrfToken: {
-        name: `${useSecureCookies ? '__Host' : ''}next-auth.csrf-token`,
-        options: secureCookieOptions,
-      },
+    callbackUrl: {
+      name: `${cookiePrefix}next-auth.callback-url`,
+      options: secureCookieOptions,
     },
-    callbacks: {
-      async redirect({ url, baseUrl }) {
-        // Allows relative callback URLs
-        if (url.startsWith('/')) return `${baseUrl}${url}`;
-
-        // Allows callback URLs on the same origin or that are permitted via CORS settings.
-        const origin = new URL(url).origin;
-        if (ORIGIN_MATCH?.test(origin) || origin === baseUrl) {
-          return url;
-        } else {
-          return baseUrl;
-        }
-      },
-      async signIn({ user }) {
-        if (!user.email) return '/api/auth/login';
+    csrfToken: {
+      name: `${useSecureCookies ? '__Host' : ''}next-auth.csrf-token`,
+      options: secureCookieOptions,
+    },
+  },
+  callbacks: {
+    // This callback allows us to get other information about the user for use in access control.
+    async session({ session, user }) {
+      if (session.user) {
+        session.user.id = user.id;
 
-        // We want to prevent users who aren't in the system from being able to login.
-        const dbUser = await client.user.findUnique({
+        const roles = await client.userSystemRole.findMany({
           where: {
-            email: user.email,
+            userId: user.id,
           },
         });
-        if (dbUser) {
-          await new Promise((resolve) =>
-            setTimeout(resolve, Math.floor(Math.random() * 300))
-          );
-          return true;
-        } else {
-          // Since sending the login email takes some time,
-          // we don't want potential attackers to be able to use the request time to determine if a user exists or not.
-          // To mitigate this, we obfuscate the response time by adding some random delay.
-          await new Promise((resolve) =>
-            setTimeout(resolve, 1800 + Math.floor(Math.random() * 500))
-          );
-          // We redirect to the verify login page rather than showing an error message
-          // so that an attacker can't determine if the account exists.
-          return '/api/auth/verify-request';
-        }
-      },
+        session.user.systemRoles = roles.map(({ role }) => role);
```

Comment:
In order for this to work we have to load the users roles from the database in their session. This will be used both on the server for access control, and by the frontend to hide and show UI